worlds/doom_1993/__init__.py

```diff
@@ -0,0 +1,185 @@
+        "A guide to setting up the DOOM 1993 randomizer connected to an Archipelago Multiworld",
+        "English",
+        "setup_en.md",
+        "setup/en",
+        ["Daivuk"]
+    )]
+    theme = "dirt"
+
+
+class DOOM1993World(World):
+    """
+    Developed by id Software, and originally released in 1993, DOOM pioneered and popularized the first-person shooter, setting a standard for all FPS games.
+    """
+    option_definitions = Options.options
+    game = "DOOM 1993"
+    web = DOOM1993Web()
+    data_version = 0
+    required_client_version = (0, 3, 9)
+    set_rules = Rules.set_rules
+
+    item_name_to_id = {data["name"]: item_id for item_id, data in Items.item_table.items()}
+    item_name_groups = Items.item_name_groups
+    item_name_to_item:Dict[str, Item] = {}
+
+    location_name_to_id = {data["name"]: loc_id for loc_id, data in Locations.location_table.items()}
+    location_name_groups = Locations.location_name_groups
+
+    starting_items:List(str) = [
+        'Hangar',
+        'Deimos Anomaly',
+        'Hell Keep'
+    ]
+
+
+    def create_regions(self):
+        # For now, include everything. Levels could eventually be split into regions
+        self.multiworld.regions += [
+            self.create_region("Menu", None, ["Level Select"]),
+            self.create_region("Mars", Events.events + [location["name"] for location in Locations.location_table.values()])
+        ]
+
+        # Link regions
+        self.multiworld.get_entrance("Level Select", self.player).connect(self.multiworld.get_region("Mars", self.player))
+
+        # Create events
+        for event in Events.events:
+            self.multiworld.get_location(event, self.player).place_locked_item(
+                DOOM1993Item(event, ItemClassification.progression, None, player=self.player))
+    
+        # Special case for E2M6 and E3M8, where you enter a normal door then get stuck behind with a key door.
+        # We need to put the key in the locations available behind this door.
+        self.place_locked_item_in_locations("Halls of the Damned - Yellow skull key", [
+            "Halls of the Damned - Yellow skull key",
+            "Halls of the Damned - Partial invisibility 2"
+        ])
+        self.place_locked_item_in_locations("Warrens - Blue skull key", [
+            "Warrens - Rocket launcher",
+            "Warrens - Rocket launcher 2",
+            "Warrens - Partial invisibility",
+            "Warrens - Invulnerability",
+            "Warrens - Supercharge",
+            "Warrens - Berserk",
+            "Warrens - Chaingun"
+        ])
+    
+
+    def place_locked_item_in_locations(self, item_name, locations):
+        location = self.multiworld.random.choice(locations)
+        self.multiworld.get_location(location, self.player).place_locked_item(self.create_item(item_name))
+
+
+    def generate_early(self):
+        # Give starting levels right away
+        for starting_item in self.starting_items:
```

Comment:
i think this should also be inside `def create_items(self) -> None:` as per #1460 but i am not 100% sure on this one